### DISCUSSION TITLE
Py03 and Socketcan CanFD Module "No such device"

### DISCUSSION THREAD

**Discussion by dulouie** [PyO3/pyo3#4556] [Questions]
Hi all, I want to develop a fast python module that can send CanFD messages using socketcan, here is the code I came up with:

```rust
use embedded_can::{Frame, StandardId};
use pyo3::prelude::*;
use pyo3::types::PyList;
use socketcan::{CanFdFrame, CanFdSocket, Socket};
use std::{env, thread, time};

#[pyfunction]
fn send(py: Python<'_>, interface: String, data: Py<PyList>) -> PyResult<()> {
    let iface = env::args().
nth
(1).unwrap_or_else(|| interface.into());
    let socket_tx = CanFdSocket::open(&iface).unwrap();
    let id = StandardId::new(0x7F0).unwrap();
    let data = data.extract::<Vec<u8>>(py).unwrap();
    let frame = CanFdFrame::new(id, &data).unwrap();
    let ten_millis = time::Duration::from_millis(10);

    loop {
        println!("Writing on {}", iface);
        socket_tx.write_frame(&frame)?;
        py.check_signals()?;
        thread::sleep(ten_millis);
    }
}

#[pymodule]
fn rust_canfd(m: &Bound<'_, PyModule>) -> PyResult<()> {
    m.add_function(wrap_pyfunction!(send, m)?)?;
    Ok(())
}
```

When I run it in bash using the pyo3 virtual environment with "maturin develop" and then the python interactive REPL it runs very well, like this:
`python`
`import rust_canfd`
`rust_canfd.send("can1",[1,1])`

But when i run it from a python file.py and not the interactive REPL it does not work and throws the following exception.

`/home/pi/.pyenv/versions/pyo3/bin/python /home/pi/src/rust_canfd/src/test.pythread '<unnamed>' panicked at src/lib.rs:11:47:called\Result::unwrap()` on an `Err` value: Os { code: 19, kind: Uncategorized, message: "No such device" }`note: run with \RUST_BACKTRACE=1` environment variable to display a backtrace`Traceback (most recent call last):File "/home/pi/src/rust_canfd/src/test.py", line 70, in <module>rust_canfd.send("can1", data)pyo3_runtime.PanicException: called \Result::unwrap()` on an `Err` value: Os { code: 19, kind: Uncategorized, message: "No such device" }``

Can someone help me?

**Comment by birkenfeld** [comment]
First, I'm not sure that we can help here, this looks like a failure in the underlying `socketcan` library opening some CAN socket/device, which is not related to the Rust/Python interface.

However, it's not clear because the output you posted is garbled and very hard to understand. Please also activate `RUST_BACKTRACE=1` as suggested.
